Sources/Apodini/Semantic Model Builder/InterfaceExporter/InterfaceExporter.swift

```diff
@@ -0,0 +1,98 @@
+protocol WithEventLoop {
+    var eventLoop: EventLoop { get }
+}
+
+
+/// Any Apodini Interface Exporter must conform to this protocol.
+protocol InterfaceExporter {
+    /// Defines the type of the Request the exporter uses.
+    associatedtype ExporterRequest: Apodini.ExporterRequest
+    /// Defines the return type of the `export` method. The return type is currently unused.
+    associatedtype EndpointExportOutput = Void
+    /// Defines the return ype of the `exportParameter` method. For more details see `exportParameter(...)`
+    associatedtype ParameterExportOutput = Void
+
+    init(_ app: Application)
+
+    /// This method is called for every `Endpoint` on start up, which must be exporter
+    /// by the `InterfaceExporter`.
+    ///
+    /// - Parameter endpoint: The `Endpoint` which is to be exported.
+    /// - Returns: `EndpointExportOutput` which is defined by the `InterfaceExporter`.
+    func export<C: Component>(_ endpoint: Endpoint<C>) -> EndpointExportOutput
+
+    /// This optional method can be defined to export a `EndpointParameter`.
+    /// It is called for every `EndpointParameter` on an `Endpoint` when calling `Endpoint.exportParameters(...)`.
+    /// `Endpoint.exportParameters(...)` returns an array of `ParameterExportOutput`, being what is returned by this method.
+    ///
+    /// - Parameter parameter:
+    /// - Returns: `ParameterExportOutput` which is defined by the `InterfaceExporter`
+    func exportParameter<Type: Codable>(_ parameter: EndpointParameter<Type>) -> ParameterExportOutput
+
+    /// This method is called once all `Endpoint`s are exported, meaning after `export` was called
+    /// for every `Endpoint` on the `WebService`.
+    /// A `InterfaceExporter` is not required to implement that method.
+    ///
+    /// - Parameter webService: A model representing the exported `WebService`
+    func finishedExporting(_ webService: WebServiceModel)
+
+    /// This method is called on `EndpointParameter` injection to retrieve the value from the given `ExporterRequest`.
+    /// Be aware that the generic `Type` holds the Wrapped type for `Optional`s (see `EndpointParameter`).
+    ///
+    /// If the value couldn't be found on the `ExporterRequest` for the given `EndpointParameter`, the `interfaceExporter`
+    /// must return `nil`. Checking for required parameters is done automatically.
+    ///
+    /// - Parameters:
+    ///   - parameter: The `EndpointParameter` describing the parameter for which the value should be retrieved.
+    ///   - request: The `ExporterRequest` as defined by the `InterfaceExporter`
+    /// - Returns: The retrieved value or nil if `ExporterRequest` didn't contain a value for the given `EndpointParameter`.
+    /// - Throws: Any Apodini Error or any other error happening while decoding.
+    func retrieveParameter<Type: Decodable>(_ parameter: EndpointParameter<Type>, for request: ExporterRequest) throws -> Type?
```

Comment:
As we discussed (also in context of [this](https://github.com/Apodini/Apodini/issues/80#issuecomment-751294746) message on #80), the exporter needs to know if an explicit `null` is a valid option. At least this holds true if we want to enable exporters to do parameter validation (which I would argue to be a reasonable approach until the shared semantic model is more mature).

This information could be provided via a property (enum/bool) on `EndpointParameter` which signalizes if `Type` was unwrapped from an `Optional<Type>`.

In any case the exporter must be able to express in the `retrieveParameter` if the value is `nil` (i.e. not present -> does not override default or previous value in client-stream) or `null` (i.e. explicitly null -> does override). I would suggest the following solution (return type is changed to `Optional<Optional<Type>>`:
```suggestion
    func retrieveParameter<Type: Decodable>(_ parameter: EndpointParameter<Type>, for request: ExporterRequest) throws -> Type??
```

Where .some(nil) would be the explicit null. Exporters that don't differentiate _not present_ and _null_ return `nil` or an instance of `Type` as before. Only exporters that make this differentiation have to take care about the **null** `.some(nil)`-case. The special meaning could be stressed using an extension on `Optional`:
```swift
extension Optional where Wrapped: ExpressibleByNilLiteral {
    static var null: Self {
        .some(nil)
    }
}
```

Reply:
Resolved via 98051ce